src/pose-sensor.js

```diff
@@ -0,0 +1,170 @@
+
+// Frequency which the Sensors will attempt to fire their
+// `reading` functions at. Use 16.6ms to achieve 60FPS, since
+// we generally can't get higher without native VR hardware
+const SENSOR_FREQUENCY = 16.6;
+
+const X_AXIS = new Vector3(1, 0, 0);
+const Z_AXIS = new Vector3(0, 0, 1);
+
+let orientation = {};
+if (screen.orientation) {
+  orientation = screen.orientation;
+} else if (screen.msOrientation) {
+  orientation = screen.msOrientation;
+} else {
+  Object.defineProperty(orientation, "angle", {
+    get: () => { return (window.orientation || 0) }
+  });
+}
+
+// Quaternion to rotate from sensor coordinates to WebVR coordinates
+const SENSOR_TO_VR = new Quaternion();
+SENSOR_TO_VR.setFromAxisAngle(X_AXIS, -Math.PI / 2);
+SENSOR_TO_VR.multiply(new Quaternion().setFromAxisAngle(Z_AXIS, Math.PI / 2));
+
+/**
+ * An abstraction class around either using the new RelativeOrientationSensor,
+ * or `devicemotion` events with complimentary filter via fusion-pose-sensor.js.
+ */
+export default class PoseSensor {
+  constructor(config) {
+    this.config = config;
+    this.sensor = null;
+    this.fusionSensor = null;
+    this._out = new Float32Array(4);
+
+    // Can be 'sensor' (using RelativeOrientationSensor) or
+    // 'devicemotion' (using devicemotion events via FusionPoseSensor)
+    this.api = 'sensor';
+
+    // Store any errors from Sensors
+    this.error = null;
+
+    // Quaternions for caching transforms
+    this._sensorQ = new Quaternion();
+    this._worldToScreenQ = new Quaternion();
+    this._outQ = new Quaternion();
+
+    this._onSensorRead = this._onSensorRead.bind(this);
+    this._onSensorError = this._onSensorError.bind(this);
+    this._onOrientationChange = this._onOrientationChange.bind(this);
+
+    this._onOrientationChange();
+    this.init();
+  }
+
+  init() {
+    // Attempt to use the RelativeOrientationSensor from Generic Sensor APIs.
+    // First available in Chrome M63, this can fail for several reasons, and attempt
+    // to fallback to devicemotion. Failure scenarios include:
+    //
+    // * Generic Sensor APIs do not exist, fallback to devicemotion
+    // * Underlying sensor does not exist, no fallback possible.
+    // * Feature policy failure (in an iframe); no fallback.
+    // * Permission to sensor data denied; respect user agent, no fallback to devicemotion.
+    //   Browsers are heading towards disabling devicemotion when sensors are denied as well.
+    let sensor = null;
+    try {
+      sensor = new RelativeOrientationSensor({ frequency: SENSOR_FREQUENCY });
+      sensor.addEventListener('error', this._onSensorError);
+    } catch (error) {
+      this.error = error;
+
+      if (error.name === 'SecurityError') {
+        console.error('Cannot construct sensors due to the feature policy');
+      } else if (error.name === 'ReferenceError') {
+        // Fallback to devicemotion
+        this.api = 'devicemotion';
+        this.useDeviceMotion();
```

Comment:
Will you attempt the polyfills for the fallback case? I would be interested in how well they work and whether things can be improved.

Reply:
The polyfills are not doing manual fusion but also relying on the deviceorientation / deviceorientationabsolute events and should handle the differences between Android and iOS - at least @anssik and I tested that a lot manually.

Reply:
I'm a bit hesitant on changing the current sensor fusion code for devicemotion, there's a few edge cases in there ([landscape-primary devices](https://github.com/googlevr/cardboard-vr-display/blob/106e6cbee87bd2001045844d8e9d8d8fe7b58c42/src/sensor-fusion/fusion-pose-sensor.js#L150), browser-specific issues between Chrome vs Firefox) but would love to give it a go. Smoothing out the differences between Android & iOS is a huge win, but even on Android there are some differences between browsers. I'm also not 100% knowledgeable about the complexities in the current sensor fusion work here which also leads to some of my hesitation of changing existing behavior :smile:

Reply:
That is understandable, no rush. I would love to have those workarounds and issues fixed in the polyfill as well. I should spend some time on testing the polyfills with Firefox on Android. Do you know any other browser on Android which differs in behavior?

Well, if RelativeOrientationSensor works great, then in theory the polyfill should work the same way at least on Android and most likely on iOS (with the exception that is does the screen orientation adjustment at this moment, but I might turn that off depending on discussions).

For the rest of the browsers I cannot say anything at this point, until I get to test them, so will start with Firefox.

Reply:
There's `devicemotion`'s `rotationRate` on Chrome being different than Safari/iOS and Firefox/Android (radians vs degrees), I'll admit I'm less familiar with `deviceorientation` and not sure how it relates to `devicemotion` and if those differences across browsers affect both events. All the differences I'm aware of are just from me seeing them in the fusion-sensor-pose file and tracking down commit history of those little fixes.

The screen orientation adjustment in the sensor polyfill confused me a bit, because if I was using native RelativeOrientationSensor, I'd still have to take into account screen orientation (which the polyfill handles already) resulting in different results for `quaternion` in the native and polyfill scenarios, but maybe I'm misunderstanding!

Reply:
That's great to hear deviceorientation doesn't suffer from some of the same issues as devicemotion when it comes to different platforms -- I need to do more research on these two events and how they relate (as well as browser support on older devices).

Ah I understand know with the polyfill vs native implementation on sensor orientations; I'll follow along in that issue on the spec.

Reply:
frequency: 60 ? // it's in Hz in the native RelativeOrientationSensor implementation

Reply:
@pozdnyakov good catch, I was using seconds here (16.6) instead of hz -- will update